visualization/open3d_vis_utils.py

```diff
@@ -0,0 +1,149 @@
+"""
+Utilities for rendering camera frustums and 3d point clouds using Open3d.
+
+Author: John Lambert
+"""
+
+import argparse
+from typing import List
+
+import numpy as np
+import open3d
+from colour import Color
+from gtsam import Cal3Bundler, Pose3
+
+from gtsfm.common.view_frustum import ViewFrustum
+
+
+def create_colored_point_cloud_open3d(point_cloud: np.ndarray, rgb: np.ndarray) -> open3d.geometry.PointCloud:
+    """Render a point cloud as individual colored points, using Open3d.
+
+    Args:
+        point_cloud: array of shape (N,3) representing 3d points.
+        rgb: uint8 array of shape (N,3) representing colors in RGB order, in the range [0,255]
+
+    Returns:
+        pcd: Open3d geometry object representing a colored 3d point cloud.
+    """
+    colors = rgb.astype(np.float64) / 255
+
+    pcd = open3d.geometry.PointCloud()
+    pcd.points = open3d.utility.Vector3dVector(point_cloud)
+    pcd.colors = open3d.utility.Vector3dVector(colors)
+
+    return pcd
+
+
+def create_colored_spheres_open3d(
+    sphere_radius: float, point_cloud: np.ndarray, rgb: np.ndarray
+) -> List[open3d.geometry.TriangleMesh]:
+    """Create a colored sphere mesh for every point inside the point cloud, using Open3d.
+
+    Note: this is quite computationally expensive.
+
+    Args:
+        sphere_radius: radius of each rendered sphere.
+        point_cloud: array of shape (N,3) representing 3d points.
+        rgb: uint8 array of shape (N,3) representing colors in RGB order, in the range [0,255]
+
+    Returns:
+        spheres: list of Open3d geometry objects, where each element (a sphere) represents a 3d point.
+    """
+    colors = rgb.astype(np.float64) / 255
+
+    spheres = []
+    n = point_cloud.shape[0]
+
+    for j in range(n):
+        wTj = np.eye(4)
+        wTj[:3, 3] = point_cloud[j]
+        mesh = open3d.geometry.TriangleMesh.create_sphere(radius=sphere_radius, resolution=20)
+        mesh.transform(wTj)
+        mesh.paint_uniform_color(colors[j])
+
+        spheres.append(mesh)
+
+    return spheres
+
+
+def create_all_frustums_open3d(
+    zcwTw: Pose3, calibrations: List[Cal3Bundler], wTi_list: List[Pose3]
+) -> List[open3d.geometry.LineSet]:
+    """Render camera frustums as collections of line segments, using Open3d.
+
+    Args:
+        zcwTw: transforms world points to a new world frame where the point cloud is zero-centered
+        calibrations: calibration object for each camera
+        wTi_list: list of camera poses for each image
```

Comment:
Args order: wTi_list, calibrations, zwcTw?

Reply:
thanks, updated

Reply:
docs not updated.

Reply:
just updated it ,thanks